### PR TITLE
[#1194] Add aggregation support for TPGM.

### PR DESCRIPTION
**Pull request by p-f**
- Add min- and max-time aggregations.

fixes #1194

### CODE REVIEW

**Comment by ChrizZz110** [comment]
First comment (haven't looked through all): Good job! You can reduce the amount of classes by removing the abstract from MinTime and MaxTime and let them implement both vertex and edge interfaces. Also rename `AbstractTimeStampAggregateFunction` to `AbstractTimeAggregateFunction`

**Comment by ChrizZz110** [comment]
I think you misundestood ... e.g. let `MaxTime` implement `VertexAggregateFunction` and `EdgeAggregateFunction` -> so you do not need `MaxEdgeTime` and `MaxVertexTime` classes anymore.
Same for `MinTime`.
Also I think the interfaces `TemporalEdgeAggregateFunction` and `TemporalEdgeAggregateFunction` are not necessary at the moment ... maybe later if we integrate it into temporal grouping.

**Comment by p-f** [comment]
Ah, I see what you mean, but I think you misunderstood as well:
An `AggregateFunction` will support vertex- and edge-aggregation by default. The `Vertex`- and `EdgeAggregateFunction` interfaces are used to restrict those functions to either vertices or edges, they can not be implemented at the same time.

In this case `MaxTime` will aggregate for both vertices and edges, `MaxVertexTime` will only aggregate vertices and `MaxEdgeTime` will only aggregate edges.

**Comment by ChrizZz110** [comment]
Oh that's my fault ... sorry for confusing. ... and thanks for the explanation

**Comment by ChrizZz110** [comment]
I updated the target Branch ... think it will be the best that you merge it into yours.

File: gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/tpgm/aggregation/MaxTime.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ * elements. This function ignores the default value ({@link Long#MAX_VALUE}) and handles
+ * it the same way as {@code null}.
+ */
+public class MaxTime extends AbstractTimeAggregateFunction {
+
+  /**
+   * The property value that is considered the default value of this aggregate function.
+   * This value is ignored during aggregation.
+   */
+  private static final PropertyValue DEFAULT = PropertyValue.create(Long.MAX_VALUE);
```
Comment (ChrizZz110) [inline gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/tpgm/aggregation/MaxTime.java]: Please use the one from `TemporalElement` class. There you can unify the existing three constants to two default values for min and max

File: gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/tpgm/aggregation/MinTime.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ * elements. This function ignores the default value ({@link Long#MIN_VALUE}) and handles
+ * it the same way as {@code null}.
+ */
+public class MinTime extends AbstractTimeAggregateFunction {
+
+  /**
+   * The property value that is considered the default value of this aggregate function.
+   * This value is ignored during aggregation.
+   */
+  private static final PropertyValue DEFAULT = PropertyValue.create(Long.MIN_VALUE);
```
Comment (ChrizZz110) [inline gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/tpgm/aggregation/MinTime.java]: Please use the one from `TemporalElement` class. There you can unify the existing three constants to two default values for min and max

File: gradoop-flink/src/test/java/org/gradoop/flink/model/impl/operators/tpgm/aggregation/MinMaxTimeTest.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+  public void testAggregationFunctions() throws Exception {
+    final String keyMaxEdge = "maxEdgeTime";
+    final String keyMinEdge = "minEdgeTime";
+    final String keyMaxVertex = "maxVertexTime";
+    final String keyMinVertex = "minVertexTime";
+    TemporalGraph result = getTestGraphWithValues().aggregate(
+      new MaxEdgeTime(keyMaxEdge, temporalAttribute, field),
+      new MinEdgeTime(keyMinEdge, temporalAttribute, field),
+      new MaxVertexTime(keyMaxVertex, temporalAttribute, field),
+      new MinVertexTime(keyMinVertex, temporalAttribute, field));
```
Comment (ChrizZz110) [inline gradoop-flink/src/test/java/org/gradoop/flink/model/impl/operators/tpgm/aggregation/MinMaxTimeTest.java]: Please add MinTime and MaxTime

**Review by ChrizZz110 — CHANGES_REQUESTED** [review]

**Comment by p-f** [comment]
Done, had to add a change to the pom, see dbs-leipzig#1256.

**Review by ChrizZz110 — APPROVED** [review]

**Pull request merged**